File: game_client/Assets/Scripts/ResScripts/NPResCommon/PlayerVariable/_ANPBasicPlayerVariableObj.cs
using System;
using System.Collections.Generic;
using ALPackage;
using GOE.Variable;
using NPEnum;

namespace GOE
{
    /*******************
     * 变量结构体基类
     **/
    public abstract class _ANPBasicPlayerVariableObj : _ITNPBasicVariable<ENPPlayerVariableType
#if NP_GAME
            , NPPlayer
#endif
        >
    {
        /******************
         * 获取条件类型
         */
        public abstract ENPPlayerVariableType variableType { get; }

        public abstract long calPlayerValue(NPVarInfo _variableInfo);

#if NP_GAME
        /// <summary>
        /// 判断条件是否匹配
        /// </summary>
        /// <param name="_cond"></param>
        /// <param name="_data"></param>
        /// <param name="_varVariableInfo"></param>
        /// <returns></returns>
        public long calValue(NPPlayer _data, NPVarInfo _varVariableInfo)
        {
            return calPlayerValue(_varVariableInfo);
        }
#endif


        /**************
         * 将字符串转化为本对象
         **/
        public static _ANPBasicPlayerVariableObj readVariable(string _str)
        {
            ALStringReader stringReader = new ALStringReader(_str);
            string varTypeS = stringReader.readItem('@');

            if (null == varTypeS)
            {
                ALLog.Warning("空的公式字符串:" + _str);
                return null;
            }

            ENPPlayerVariableType varType = ENPPlayerVariableType.NONE;
            try
            {
                //读取类型枚举
                varType = (ENPPlayerVariableType)ALCommon.EnumParse(typeof(ENPPlayerVariableType), varTypeS, true);
            }
            catch (Exception) { }

            if (varType == ENPPlayerVariableType.NONE)
            {
#if UNITY_EDITOR
                ALLog.Error("错误的公式类型:" + _str);
#endif
                return null;
            }

            return readVariable(varType, stringReader);
        }

        /********************
         * 从节点中读取相关信息
         */
        public static _ANPBasicPlayerVariableObj readVariable(ENPPlayerVariableType _variableType, ALStringReader _reader)
        {
            switch (_variableType)
            {
                case ENPPlayerVariableType.C_RND:
                    return NPPlayerVariableRnd.readVariable(_reader);

                case ENPPlayerVariableType.CS_NUM:
                    return NPPlayerVariableNum.readVariable(_reader);
                case ENPPlayerVariableType.CS_VALUE:
                    return NPPlayerVariablePlayerValue.readVariable(_reader);
                case ENPPlayerVariableType.CS_PROPERTY:
                    return NPPlayerVariablePlayerPro.readVariable(_reader);
                case ENPPlayerVariableType.CS_VAR_V:
                    return NPPlayerVariableVarValue.readVariable(_reader);
                case ENPPlayerVariableType.CS_BUF_L:
                    return NPPlayerVariableBufL.readVariable(_reader);
                case ENPPlayerVariableType.CS_BUFF_HAD_ACTIVE_DAY:
                    return NPPlayerVariable_CS_BUFF_HAD_ACTIVE_DAY.readVariable(_reader);
                case ENPPlayerVariableType.CS_ITEM_COUNT:
                    return NPPlayerVariableItemCount.readVariable(_reader);

                case ENPPlayerVariableType.CS_PARAM:
                    return NPPlayerVariablePlayerParam.readVariable(_reader);
                case ENPPlayerVariableType.CS_VALUE_ID:
                    return NPPlayerVariableIdValue.readVariable(_reader);
                // case ENPPlayerVariableType.CS_PET_NUM_L:
                //     return NPPlayerVariablePetNumL.readVariable(_reader);
                // case ENPPlayerVariableType.CS_PET_NUM_S:
                //     return NPPlayerVariablePetNumS.readVariable(_reader);
                // case ENPPlayerVariableType.CS_PET_NUM_E:
                //     return NPPlayerVariablePetNumE.readVariable(_reader);
                case ENPPlayerVariableType.CS_CONDITION_BOOLEAN_V:
                    return NPPlayerVariableConditionBooleanV.readVariable(_reader);
                case ENPPlayerVariableType.CS_EVENT_RECORD:
                    return NPPlayerVariableEventRecord.readVariable(_reader);
                case ENPPlayerVariableType.CS_SCOPE_ITEM_COUNT:
                    return NPPlayerVariableScopeItemCount.readVariable(_reader);
                case ENPPlayerVariableType.CS_HERO_NUM:
                    return NPPlayerVariableHeroNum.readVariable(_reader);
                case ENPPlayerVariableType.CS_RECORD_PARAM:
                    return NPPlayerVariableRecordParam.readVariable(_reader);
                case ENPPlayerVariableType.CS_CONSORT_RES_COUNT:
                    return NPPlayerVariable_CS_CONSORT_RES_COUNT.readVariable(_reader);
                case ENPPlayerVariableType.CS_BUILDING_LEVEL:
                    return NPPlayerVariable_CS_BUILDING_LEVEL.readVariable(_reader);
                case ENPPlayerVariableType.CS_HAS_BUILDING:
                    return NPPlayerVariable_CS_HAS_BUILDING.readVariable(_reader);
                case ENPPlayerVariableType.CS_BUSINESS_BUILDING_WORKER_NUM:
                    return NPPlayerVariable_CS_BUSINESS_BUILDING_WORKER_NUM.readVariable(_reader);
                case ENPPlayerVariableType.CS_CONSORT_FETTER_LEVLE:
                    return NPPlayerVariable_CS_CONSORT_FETTER_LEVLE.readVariable(_reader);
                case ENPPlayerVariableType.CS_MARS_BUILDING_LVL:
                    return NPPlayerVariable_CS_MARS_BUILDING_LVL.readVariable(_reader);
                case ENPPlayerVariableType.CS_MARS_BUILDING_EQUIP_LVL:
                    return NPPlayerVariable_CS_MARS_BUILDING_EQUIP_LVL.readVariable(_reader);
                case ENPPlayerVariableType.CS_MARS_BUILDING_DISPATCH_NUM:
                    return NPPlayerVariable_CS_MARS_BUILDING_DISPATCH_NUM.readVariable(_reader);
                default:
                    return null;
            }
        }
    }
}

File: game_server/GameRes/src/NPGameRes/GameObjs/PlayerVariable/_ANPBasicPlayerVariableObj.java
package NPGameRes.GameObjs.PlayerVariable;

import ALServerLog.ALServerLog;
import NPCommon.CommonObj.NPStringReader;
import NPEnum.ENPPlayerVariableType;
import NPGameRes.GameObjs.CommonObj.Variable.InterfaceObj._ITNPBasicVariable;
import NPGameRes.GameObjs.PlayerVariable.PlayerVariableObj.*;

/*******************
 * 玩家参数信息的基类
 * @author mj
 *
 */
public abstract class _ANPBasicPlayerVariableObj implements _ITNPBasicVariable<ENPPlayerVariableType>
{
    /**************
     * 将字符串转化为本对象
     **/
    public static _ANPBasicPlayerVariableObj readVariable(String _str)
    {
        NPStringReader stringReader = new NPStringReader(_str);

        String vTypeS = stringReader.readItem('@');

        if (null == vTypeS)
        {
            ALServerLog.Error("Empty Variable Type: " + _str);
            return null;
        }

        //读取类型枚举
        ENPPlayerVariableType varType = ENPPlayerVariableType.NONE;

        try
        {
            varType = ENPPlayerVariableType.valueOf(vTypeS.toUpperCase());
        } catch (Exception e)
        {
        }

        if (varType == ENPPlayerVariableType.NONE)
        {
            ALServerLog.Error("Empty Variable Type: " + _str);
            return null;
        }

        return _readVariable(varType, stringReader);
    }

    /********************
     * 从节点中读取相关信息
     */
    private static _ANPBasicPlayerVariableObj _readVariable(ENPPlayerVariableType _variableType, NPStringReader _reader)
    {
        switch (_variableType)
        {
            case S_RND:
                return NPPlayerVariable_S_RND.readVariable(_reader);
            case CS_NUM:
                return NPPlayerVariable_CS_NUM.readVariable(_reader);
            case CS_VALUE:
                return NPPlayerVariable_CS_VALUE.readVariable(_reader);
            case CS_PROPERTY:
                return NPPlayerVariable_CS_PROPERTY.readVariable(_reader);
            case CS_BUF_L:
                return NPPlayerVariable_CS_BUF_L.readVariable(_reader);
            case CS_BUFF_HAD_ACTIVE_DAY:
                return NPPlayerVariable_CS_BUFF_HAD_ACTIVE_DAY.readVariable(_reader);
            case CS_ITEM_COUNT:
                return NPPlayerVariable_CS_ITEM_COUNT.readVariable(_reader);
            case CS_PARAM:
                return NPPlayerVariable_CS_PARAM.readVariable(_reader);
            case CS_VALUE_ID:
                return NPPlayerVariable_CS_VALUE_ID.readVariable(_reader);
            case CS_VAR_V:
                return NPPlayerVariable_CS_VAR_V.readVariable(_reader);
            case CS_CONDITION_BOOLEAN_V:
                return NPPlayerVariable_CS_CONDITION_BOOLEAN_V.readVariable(_reader);
            case CS_RECORD_PARAM:
                return NPPlayerVariable_CS_RECORD_PARAM.readVariable(_reader);
            case CS_EVENT_RECORD:
                return NPPlayerVariable_CS_EVENT_RECORD.readVariable(_reader);
            case CS_SCOPE_ITEM_COUNT:
                return NPPlayerVariable_CS_SCOPE_ITEM_COUNT.readVariable(_reader);
            case S_RND_ATTR_HERO:
                return NPPlayerVariable_S_RND_ATTR_HERO.readVariable(_reader);
            case S_RND_FROM_SCOPE:
                return NPPlayerVariable_S_RND_FROM_SCOPE.readVariable(_reader);
            case S_RND_CONSORT:
                return NPPlayerVariable_S_RND_CONSORT.readVariable();
            case CS_HERO_NUM:
                return PlayerVariable_CS_HERO_NUM.readVariable(_reader);
            case S_RND_VALUE_BY_WEIGHT:
                return PlayerVariable_S_RND_VALUE_BY_WEIGHT.readVariable(_reader);
            case CS_LEVY_SUM:
            	return PlayerVariable_CS_LEVY_SUM.readVariable(_reader);
            case CS_CONSORT_RES_COUNT:
            	return PlayerVariable_CS_CONSORT_RES_COUNT.readVariable(_reader);
            case CS_HAS_BUILDING:
            	return PlayerVariable_CS_HAS_BUILDING.readVariable(_reader);
            case CS_BUSINESS_BUILDING_WORKER_NUM:
            	return PlayerVariable_CS_BUSINESS_BUILDING_WORKER_NUM.readVariable(_reader);
            case CS_BUILDING_LEVEL:
            	return PlayerVariable_CS_BUILDING_LEVEL.readVariable(_reader);
            case CS_MARS_BUILDING_LVL:
            	return NPPlayerVariable_CS_MARS_BUILDING_LVL.readVariable(_reader);
            case CS_MARS_BUILDING_EQUIP_LVL:
            	return NPPlayerVariable_CS_MARS_BUILDING_EQUIP_LVL.readVariable(_reader);
            case CS_MARS_BUILDING_DISPATCH_NUM:
            	return NPPlayerVariable_CS_MARS_BUILDING_DISPATCH_NUM.readVariable(_reader);
            default:
                return null;
        }
    }
}
